In [ ]:
import txtai

In [ ]:
txtai.__version__

'5.0.0'

Install prerequisite packages.

# YouTube Indexing and Queries

In this notebook we will work through an example of indexing and querying the YouTube video transcriptions data. We start by loading the dataset.

In [ ]:
from datasets import load_dataset
import pandas as pd

In [ ]:
ytt = load_dataset(
    "./data",
    split="train",
)
ytt

2022-09-30 23:14:56,487 [WARNING] _create_builder_config: Using custom data configuration data-02fffa30d1cc87c7


Extracting data files:   0%|          | 0/1 [00:00<?, ?it/s]

0 tables [00:00, ? tables/s]

Dataset json downloaded and prepared to /home/matu/.cache/huggingface/datasets/json/data-02fffa30d1cc87c7/0.0.0/e6070c77f18f01a5ad4551a8b7edfba20b8438b7cad4d94e6ad9378022ce4aab. Subsequent calls will reuse this data.


Dataset({
    features: ['url', 'title', 'text', 'start_second', 'end_second'],
    num_rows: 8806
})

Each sample includes video-level information (ID, title, url and thumbnail) and snippet-level information (text, start_second, end_second).

In [ ]:
for x in ytt:
    print(x)
    break

{'url': 'https://www.youtube.com/watch?v=8SF_h3xF3cE&t=0s', 'title': 'Lesson 1: Practical Deep Learning for Coders 2022', 'text': "  Welcome to practical deep learning for coders lesson one.  This is version five of this course.  And it's the first do one we've done in two years.  So we've got a lot of cool things to cover.  It's amazing how much has changed.  Here is a XKCD from the end of 2015.  Who here is saying XKCD comics before?  Pretty much everybody, not surprising.  So the basic joke here is I'll let you read it,  and then I'll come back to it.  So it can be hard to tell what's easy and what's nearly impossible.", 'start_second': 0.0, 'end_second': 58.0}


# generate the TXTAI database with vectors

The next step is indexing this dataset in Pinecone, for this we need a sentence transformer model (to encode the text into embeddings), and a Pinecone index.

We will initialize the sentence transformer first.

In [ ]:
from txtai.embeddings import Embeddings

In [ ]:
embeddings = Embeddings(
                        {
                            "path": "sentence-transformers/multi-qa-mpnet-base-dot-v1",
                            "content": True,
                            "objects": True
                        }
                       )

In [ ]:
embeddings.index([(uid, data, None) for uid, data in enumerate(ytt)])

In [ ]:
pd.DataFrame(embeddings.search('what is a convolutional neural network?',limit=50)).sort_values('score')

id  \
49  5707   
48  3331   
47  8424   
46  1316   
45  2750   
44  2760   
43  8410   
42  1322   
41  4195   
40  1323   
39  2754   
38  5897   
37  4196   
36  5807   
35  1300   
34  8480   
33  2753   
32  1303   
31  8479   
30  8581   
29  5842   
28  8580   
27  8449   
26  1326   
25  2751   
24  2752   
23  8478   
22  3876   
21   438   
20  1299   
19   439   
18  2759   
17  1320   
16  8707   
15  1309   
14  5843   
13  8477   
12  1310   
11  1302   
10  1305   
9   2755   
8   2758   
7   1304   
6   2756   
5   2757   
4   1306   
3   8411   
2   8412   
1   8414   
0   8413   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   text  \
49    Let's more discuss in the next lesson, we'll talk a lot about neural net architecture details.  But the details we'll focus on are what happens to the inputs at the very first stage and  what happens to the outputs at the very last stage.  We'll talk a bit about what happens in the middle, but a lot less.  And the reason why is it's a thing that you put into the inputs that's going to change  for every single data set you do.  And what do you want to happen to the outputs that are going to...   
48    It's modifying things together and adding them up.  So there'd be one more step to do to make this a layer of a neural network which is  if this had any negatives, we'd place them with yours.  That's why matrix multiplication is the critical foundation or mathematical operation in basically  all of deep learning.  So the GPUs that we use, the thing that they are good at is this, matrix multiplication.  They have special cores called tensor cores which can basically only do one thing which ...   
47    And what you see in the in the next column is a version of the image where the horizontal lines being recognized.  And another one where the vertical lines are being recognized. And if you think back to that Xyla and Fergus paper that talked about what the layers of a neural net does, this is an absolutely an example of something that we know that the first layer of a neural network  tends to learn how to do.  Now, how did I do this? I did this using something called a convolution.  And so...   
46                             And then it takes those as inputs to a next layer.  It does the same thing.  It multiplies them a bunch of times and adds them up.  And it does that a few times.  And that's called a neural network.  Now, the model, therefore, is not  going to do anything useful.  And this leads weights to very carefully, it shows it.  And so the way it works is that we actually  start out with these weights as being random.  So initially, this thing doesn't do anything useful at all.   
45    We could have a look at zero dot model dot stages dot zero dot blocks dot one dot MLP  dot F C one and parameters and other big bunch of numbers.  So what's going on here?  What are these numbers and where it is that they come from and how come these numbers  can figure out whether something is a basset hound or not?  Okay.  So to answer that question, we're going to have a look at a Kaggle notebook.  How does a neural network really work?  I've got a local version of it here, which I'm go...   
44           So we start out with a very, very flexible, in fact, an infinitely flexible as we discussed  function, a neural network.  And we get it to do a particular thing, which is to recognize the patterns in the data examples  we give it.  So let's do a much simpler example than a neural network.  Let's do a quadratic.  So let's create a functi

In [ ]:
ytt['url'][8413]

'https://www.youtube.com/watch?v=htiNBPxcXgo&t=2757s'

In [ ]:
embeddings.save()

In [ ]:
https://www.youtube.com/watch?v=8SF_h3xF3cE&list=PLfYUBJiXbdtSvpQjSnJJ_PmDQB_VyT5iU&index=1&t=1713s

In [ ]:
'https://www.youtube.com/watch?v=8SF_h3xF3cE&t=1713s'

In [ ]:
from sentence_transformers import SentenceTransformer

retriever = SentenceTransformer('flax-sentence-embeddings/all_datasets_v3_mpnet-base')
retriever

SentenceTransformer(
  (0): Transformer({'max_seq_length': 128, 'do_lower_case': False}) with Transformer model: MPNetModel 
  (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False})
  (2): Normalize()
)

We can see the embedding dimension of `768` above, we will need this when creating our Pinecone index.

In [ ]:
embed_dim = retriever.get_sentence_embedding_dimension()
embed_dim

768

Now we can initialize our index.

In [ ]:
index.describe_index_stats()

{'dimension': 768,
 'index_fullness': 0.01,
 'namespaces': {'': {'vector_count': 11298}}}

# Querying

When query we encode our text with the same retriever model and pass it to the Pinecone `query` endpoint.

In [ ]:
pd.set_option('display.max_colwidth',500)

In [ ]:
query = "What is deep learning?"

xq = retriever.encode([query]).tolist()

In [ ]:
xc = index.query(xq, top_k=5,
                 include_metadata=True)
for context in xc['results'][0]['matches']:
    print(context['metadata']['text'], end="\n---\n")

 terms of optimization but what's the algorithm for updating the parameters or updating whatever the state of the network is and then the the last part is the the data set like how do you actually represent the world as it comes into your machine learning system so I think of deep learning as telling us something about what does the model look like and basically to qualify as deep I
---
 any theoretical components any theoretical things that you need to understand about deep learning can be sick later for that link again just watched the word doc file again in that I mentioned the link also the second channel is my channel because deep learning might be complete deep learning playlist that I have created is completely in order okay to the other
---
 under a rock for the last few years you have heard of the deep networks and how they have revolutionised computer vision and kind of the standard classic way of doing this is it's basically a classic supervised learning problem you are givi